device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/AndroidEmulatorsNode.kt

```diff
@@ -0,0 +1,352 @@
+package com.badoo.automation.deviceserver.host
+
+import com.badoo.automation.deviceserver.data.*
+import com.badoo.automation.deviceserver.host.management.errors.DeviceCreationException
+import com.badoo.automation.deviceserver.host.management.errors.DeviceNotFoundException
+import com.badoo.automation.deviceserver.host.management.errors.OverCapacityException
+import okhttp3.*
+import org.slf4j.LoggerFactory
+import java.net.HttpURLConnection
+import java.net.URI
+import java.net.URL
+import java.util.*
+import java.util.concurrent.*
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.RuntimeException
+
+typealias EmulatorSerial = String
+
+// Represents one ticket-machine = ticketHost = machine:8088.
+class AndroidEmulatorsNode(
+        private val ticketHost: String,
+        private val emulatorLimit: Int,
+        private val concurrentBoots: Int,
+        private val concurrentBooter: ExecutorService = Executors.newFixedThreadPool(concurrentBoots),
+        private val httpClient: OkHttpClient = OkHttpClient()
+) : ISimulatorsNode {
+    // Unique number for references
+    private var serial = AtomicInteger(0)
+
+    // Nicer to keep a list of known serial numbers than write a validating function, because it's similar
+    // for real devices which cannot have a validating function and may disappear if its battery dies
+    // or someone borrows one.
+    private val knownEmulators = (1..emulatorLimit).map { "emulator-${5552 + 2 * it}" }.toSet()
+
+    // Need to keep a list of allocated devices; keep ticket for handy list of existing devices.
+    private val allocatedEmulators = ConcurrentHashMap<EmulatorSerial, DeviceRef?>()
```

Comment:
I don't understand why `allocatedEmulators` will contain `null` values. I think it should be `<EmulatorSerial, DeviceRef>`.

Reply:
If it's not allocated?

Reply:
Not sure either. Changed.